File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
    if(lists.size()==0)return NULL;
    priority_queue<int> pq;
    for (int i=0;i<lists.size();i++)
    {
        ListNode* temp=lists[i];
        while(temp!=NULL)
         {
            pq.push(temp->val*(-1));
            temp=temp->next;
        }
    }
    if(pq.size()==0) return NULL;
    ListNode* head=new ListNode;
    head->val=(pq.top() * (-1));
    head->next=NULL;
    pq.pop();
    
    ListNode* ans=head;
    while(!pq.empty())
    {
        ListNode* neksht=new ListNode;
        neksht->val=(pq.top() * (-1));
        neksht->next=NULL;
        ans->next=neksht;
        pq.pop();
        ans=ans->next;
        
    }
    return head;
        }
    };

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(int i=0;i<lists.length;i++){
            while(lists[i]!=null){
                pq.add(lists[i].val);
                lists[i]=lists[i].next;
            }
        }
        if (pq.isEmpty()) {
            return null;
        }
        ListNode res=new ListNode(pq.poll());
        ListNode temp=res;
        while(!pq.isEmpty()){
            ListNode newNode=new ListNode(pq.poll());
            temp.next=newNode;
            temp=temp.next;
        }
        return res;
    }
}